### PR TITLE
Remove `CombinatorialClass`

### LINKED ISSUE #12913: Meta-ticket: Deprecate CombinatorialClass in favor of the EnumeratedSet's categories
<div id="comment:0"></div>

All classes that currently inherit from `CombinatorialClass` should instead inherit from Parent and register themselves in one of the categories (`*EnumeratedSets`, `*FiniteEnumeratedSets`, or `*InfiniteEnumeratedSets`). For examples, see e.g. `FiniteEnumeratedSets().example()`.

* Expected benefits:
  * Uniform interface
  * Support for TestSuite
  * Support for conversions, coercions, and morphisms, in particular for bijections (as morphisms in the category of Sets with bijection, with properly defined domain and co-domain rather than python functions).
* Steps:
  1. Let `CombinatorialClass` inherits from `Parent` (mostly done #8910)
  2. For each `CombinatorialClass` C:
    * Have C inherit directly from `Parent`
    * Ensure that `C.__init__` sets up the proper category (Finite|Infinite)...
    * Add `TestSuite(C).run()` to the doctests and make all the tests pass
    * Have a properly setup attribute `C.Element` and use `C.element_class` (as defined by the categories) when constructing elements
    * Ensure proper unique representation behavior by having C inherit both from `UniqueRepresentation` and `Parent`
  3. Deprecate and remove `CombinatorialClass`
  4. Turn all the factory functions into factory classes by mean of `ClasscallMetaClass`; see `PerfectMatching` and `Trees` for an advanced example.

See also: http://wiki.sagemath.org/CombinatorialClass

CC:  @sagetrac-sage-combinat

Component: **combinatorics**

_Issue created by migration from https://trac.sagemath.org/ticket/12913_

### LINKED ISSUE #19474: EnumeratedSets.ParentMethods.map should not rely on CombinatorialClass
<div id="comment:0"></div>

See the `map` method in `src/sage/categories/enumerated_sets.py`. We want to deprecate combinatorial classes, so let's not use them here.

Might be a good idea to rename/alias this as `direct_image` and make this a functorial construction that works for various other categories.

CC:  @nthiery @tscrim @stumpc5

Component: **categories**

Keywords: **enumerated sets, OOP**

_Issue created by migration from https://trac.sagemath.org/ticket/19474_

### LINKED ISSUE #33384: Deprecate global imports of CombinatorialObject, CombinatorialClass, MapCombinatorialClass
<div id="comment:0"></div>

Part of #12913 (Meta-ticket: Deprecate `CombinatorialClass` in favor of the `EnumeratedSet`'s categories)

CC:  @tscrim @fchapoton

Component: **combinatorics**

Author: **Matthias Koeppe**

Branch/Commit: **[u/mkoeppe/deprecate_global_imports_of_combinatorialobject__combinatorialclass__mapcombinatorialclass](https://github.com/sagemath/sagetrac-mirror/tree/u/mkoeppe/deprecate_global_imports_of_combinatorialobject__combinatorialclass__mapcombinatorialclass) @ [`1d3a09b`](https://github.com/sagemath/sagetrac-mirror/commit/1d3a09baf94a54112382f06acacb098c9f186d27)**

_Issue created by migration from https://trac.sagemath.org/ticket/33384_

### LINKED ISSUE #36133: inconsistent depreciation of `UnionCombinatorialClass` and `FilteredCombinatorialClass`
@fchapoton This deprecation in #31545 is inconsistent. While `InfiniteAbstractCombinatorialClass` can indeed go (can be just removed now, as ~2 years passed  since its deprecation), `UnionCombinatorialClass` and `FilteredCombinatorialClass` are used in `CombinatorialClass` - which is not formally deprecated (there is a comment there saying it is, but no deprecation notice).

_Originally posted by @dimpase in https://github.com/sagemath/sage/issues/31545#issuecomment-1691357694_

### PR DESCRIPTION
**Pull request by mkoeppe**
- Fixes #36133 
- Fixes #33384
- Fixes #19474
- Fixes #12913 

### :memo: Checklist

- [ ] The title is concise and informative.
- [ ] The description explains in detail what this PR is about.
- [x] I have linked a relevant issue or discussion.
- [ ] I have created tests covering the changes.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by mkoeppe** [comment]
Ready for review.

The failure in "Test modularized distributions" is unrelated; it is tracked in #37734.

File: src/sage/combinat/free_module.py
Comment (mantepse) [inline src/sage/combinat/free_module.py]: why this change?  `is_injective` is the default, and also what I'd assume that `map` does.

File: src/sage/combinat/root_system/root_lattice_realizations.py
Comment (mantepse) [inline src/sage/combinat/root_system/root_lattice_realizations.py:1234]: why this change?  `is_injective` is the default, and also what I'd assume that `map` does.

**Comment by mantepse** [comment]
I think that's great!

**Review by tscrim — CHANGES_REQUESTED** [review]
I don't think the `ImageSubobject` should take the category into account for equality and hashing. It doesn't change what the actual set is. Same for `_is_injective`. Everything is contained within the `self._map`, `self._inverse`, and `self._domain_subset`.

Other than that, LGTM. We have to pull the trigger at some point on removing `CombinatorialClass`.

File: src/sage/combinat/free_module.py
Comment (mkoeppe) [reply src/sage/combinat/free_module.py]: just for clarity

File: src/sage/combinat/root_system/root_lattice_realizations.py
Comment (mkoeppe) [reply src/sage/combinat/root_system/root_lattice_realizations.py:1234]: just annotation for clarity as I made my way through the code to fix a failure.

**Comment by mkoeppe** [comment]
> I don't think the `ImageSubobject` should take the category into account for equality and hashing. It doesn't change what the actual set is. Same for `_is_injective`. Everything is contained within the `self._map`, `self._inverse`, and `self._domain_subset`.

Also `self._inverse` does not change what the actual set is...

File: src/sage/combinat/free_module.py
Comment (mkoeppe) [reply src/sage/combinat/free_module.py]: > also what I'd assume that `map` does.

That's one of the trickier parts in the enumerated sets code. See e.g. #34389

**Review by tscrim — APPROVED** [review]
Thank you. LGTM. @mantepse Feel free to revert if you disagree.

**Comment by mkoeppe** [comment]
Thanks!

**Comment by vbraun** [comment]
merge conflict

**Comment by mantepse** [comment]
I'm not sure what happened here, I thought it was a simple merge conflict.  Have I overlooked any changes?

**Comment by mkoeppe** [comment]
I set it to "needs review" only to wait for the CI to complete after resolving the merge conflict.
It's all good.

**Comment by vbraun** [comment]
merge conflict

**Comment by mantepse** [comment]
Hi Matthias! I suppose that you only merged develop, right?

I find it a bit hard ti review if after a merge conflict the whole branch is replaced. I would prefer to see only the difference to the last version.

**Review by mantepse — APPROVED** [review]

**Comment by mkoeppe** [comment]
> I suppose that you only merged develop, right?

No, I rebased a previous version that was based on 10.4.beta2 (2141959) onto 10.4.beta4.

**Comment by mkoeppe** [comment]
> I find it a bit hard ti review if after a merge conflict the whole branch is replaced. I would prefer to see only the difference to the last version.

That's understandable, but ease for the reviewer is only one of several criteria when deciding how to update a branch. Cleanliness of the branch is another one. 

I can teach you how to use git more effectively in such situations if you are interested.

**Comment by mantepse** [comment]
I was on my smartphone, so, no git there.

But yes, I'd still be interested to know how to get the "interesting part" of the diff between the branch that was replaced with the current branch.

**Comment by mkoeppe** [comment]
@mantepse A technique that I find useful in such situations is to take the old, reviewed version (here: `git fetch upstream 1cdc9cfb1e70a94abf149582c13d588aeaf1b43e && git checkout FETCH_HEAD`) and then merge the new release, asking for any merge conflicts to be resolved automatically instead of interactively (using `-Xtheirs` or `-Xours`): `git merge -Xtheirs upstream/develop`. Then the diff with the new version can be inspected.

**Pull request merged**
